### PR TITLE
subsys: shell: removing Console dependencies from Shell UART backend

### LINKED ISSUE #10191: The new shell uses CONSOLE kconfig options, even though it does not use CONSOLE.
The new shell uses CONSOLE kconfig options, even though it does not use CONSOLE.

Related to #10190 & #10127

### PR DESCRIPTION
**Pull request by jakub-uC**
This PR fixes #10191.

@aescolar @pfalcon @jukkar : This is my fix proposal to get rid of Console dependencies from shell uart backend. Once you will agree with this approach I will add necessary aliases for all boards.

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordicsemi.no>

### CODE REVIEW

File: boards/arm/nrf52840_pca10056/nrf52840_pca10056.dts [hunk 1]
```diff
@@ -71,6 +71,7 @@
 		sw1 = &button1;
 		sw2 = &button2;
 		sw3 = &button3;
+		config-shell-serial = &uart0;
```
Comment (carlescufi) [inline boards/arm/nrf52840_pca10056/nrf52840_pca10056.dts]: Shouldn't this be in the `chosen` section, just like `zephyr,console = &uart0;` ?

Comment (jarz-nordic) [reply boards/arm/nrf52840_pca10056/nrf52840_pca10056.dts]: I am not agains this.
Can someone point pros and cons for both solutions?

Comment (carlescufi) [reply boards/arm/nrf52840_pca10056/nrf52840_pca10056.dts]: @erwango @galak any thoughts on this?

**Review by pfalcon — DISMISSED** [review]
This is conceptually wrong. There's a hierarchy: UART -> console -> shell. Shell should not deal with UART. It currently does, and it's a bug which needs to be fixed. Not promoted to new levels with syntax like `config-shell-serial = &uart0;`, mixing "shell" and "uart" in one line.

Moreover, there's a common criteria for what belong to DT: it's hardware configuration. Which UART to use for console is not hardware configuration, I actually imagine that to be runtime settable. So, having CONFIG_UART_CONSOLE_ON_DEV_NAME is about as right as it could be (it's a bit weird name, just need to read it literally "UART console is on device named &lt;foo>".

**Review by aescolar — DISMISSED** [review]
I do not think this is solving the issue but just hiding it under several layers. I would think it should be possible to configure the backend in compile time, so it should not be hardcoded  in a board file.

**Comment by carlescufi** [comment]
@pfalcon 

>This is conceptually wrong. There's a hierarchy: UART -> console -> shell. Shell should not deal with UART. It currently does, and it's a bug which needs to be fixed. Not promoted to new levels with syntax like config-shell-serial = &uart0;, mixing "shell" and "uart" in one line.

So what you are saying is that we need to augment console with all the functionality that it is currently missing, such as multi-instance, an asynchronous API and multi-byte writes?

**Comment by pfalcon** [comment]
> This is conceptually wrong. There's a hierarchy: UART -> console -> shell. Shell should not deal with UART. It currently does, and it's a bug which needs to be fixed. 

Actually, on second thought, this seems to be a patch along the lines of  "we don't use console in shell, and let's not pretend that we do". In that regard, it's ok. 

But I agree with comments above that the setting likely belongs to `chosen` section. Generally, it should go where linux specifies default uart for (boot) console, and that's `chosen`, right? And I'm not sure it should be `zephyr,console = &uart0;` given the concerns above. If it's just for shell, when please make it `zephyr,shell-console = &uart0;`. And if it's really a global Zephyr console, then let's work together to make it such (governed by POSIX conventions).

**Comment by pfalcon** [comment]
> So what you are saying is that we need to augment console with all the functionality that it is currently missing, such as multi-instance, an asynchronous API and multi-byte writes?

Yes, that's what I've been saying, inviting all interested parties to cooperate on the evolution of subsys/console/ . More specifically, now I think it should just follow POSIX API, or we will never get out of vicious cycle of "originally designed" APIs.

But per my previous comment, I'm withdrawing my -1 here, as it indeed just clarifies the current status quo.

**Comment by pfalcon** [comment]
@carlescufi :

> So what you are saying is that we need to augment console with all the functionality that it is currently missing, such as multi-instance, an asynchronous API and multi-byte writes?

Yes, that's what I've been saying, inviting all interested parties to cooperate on the evolution of subsys/console/ . More specifically, now I think it should just follow POSIX API, or we will never get out of vicious cycle of "originally designed" APIs.

But per my previous comment, I'm withdrawing my -1 here, as it indeed just clarifies the current status quo.

**Comment by jarz-nordic** [comment]
@pfalcon : We discussed in other PR that UART -> console -> shell is something we all wish to have but current console implementation is not good enough. Conclusion was that we are welcome to improve. Question is who and when :) ?

Purpose of this PR is as you said to highlight that shell is not related with console in any way.
There is no problem at all to add console backend to the shell, but also who and when?

**Comment by jarz-nordic** [comment]
@aescolar : May you please let me know would you like to configureduring compile time? Currently we implement one shell backend called UART and UART is defined in DT.

**Comment by aescolar** [comment]
@jarz-nordic : Today, in no way.. But wasn't the idea that the shell was going to support several different backends?

**Comment by aescolar** [comment]
My point is, is there now way of doing this without modifying the 100 boards files? (and mind you, some boards do not use DTS)

**Comment by pfalcon** [comment]
@aescolar : There's DT overlays: https://docs.zephyrproject.org/latest/application/application.html#application-dt . Yes, it's less convenient than Kconfig, so would be nice to follow good criteria on what to put in DT and what not. (OTOH, it's "less convenient" because we aren't used to it. Exercising it is the way to get it better.)

**Comment by aescolar** [comment]
@jarz-nordic : I guess I missunderstood the intention of this PR. You are welcome to disregard my comments if you think they do not apply. But please, whenever you have a PR which is meant to be merged, ensure it does work also for native_posix and other boards which do not use DTS.

**Comment by pfalcon** [comment]
> but current console implementation is not good enough. Conclusion was that we are welcome to improve. Question is who and when :) ?

Well, I'm slowly (as a background activity) refactoring subsys/console to something more general, and where help is needed is in reviewing such changes. If you're interested, I can add you as a reviewer to related patches.

> Purpose of this PR is as you said to highlight that shell is not related with console in any way.

Ack, sorry for the noise, for a moment I thought it tries do bend whatever consoles we have the shell's way ;-).

> There is no problem at all to add console backend to the shell, but also who and when?

Well, as you noticed, I skipped reviewing the shell patchset (only tested external behavior), because I figured my criticism won't be productive at the current stage of affairs. So, I yet need to dig into it, try to update MicroPython code to call out to shell handlers (aka re-integrate Zephyr shell into MicroPython shell), then consider if it's possible/how to have such an integration in a generic (e.g. userspace) case. That's low-priority for now, I hope to share some insights as it goes.

**Comment by jarz-nordic** [comment]
@pfalcon : We plan to add console backend anyway quite soon. We need to have a "default" shell backend and console seems to be a good candidate.

**Comment by nordic-krch** [comment]
@pfalcon could you share what you've done so far regarding console? I feel like we have mess right now here (consoles, logger, shell). It would be good to agree on the architecture of io (console, shell, log) subystem. I will create issue where we could discuss it.

**Pull request closed without merging**